test_progs/pcq.c

```diff
@@ -0,0 +1,1066 @@
+			break;
+
+		case 't':
+			runtime = strtoull(optarg, 0, 0);
+			break;
+
+		case 'p':
+			producer = true;
+			break;
+
+		case 'c':
+			consumer = true;
+			break;
+
+		case 'd':
+			drain = true;
+			wait = false;
+			break;
+
+		case 'v':
+			verbose++;
+			break;
+
+		case 'N':
+			nmessages = strtoull(optarg, &endptr, 0);
+			mult = get_multiplier(endptr);
+			if (mult > 0)
+				nmessages *= mult;
+			break;
+			break;
+
+		case 'f':
+			/* Write execution status to this file (for testing) */
+			statusfname= optarg;
+			break;
+
+		case 'D':
+			mock_flush = 1;
+			break;
+
+		case 'h':
+		case '?':
+			pcq_usage(argc, argv);
+			return 0;
+		}
+	}
+
+	if (info && (create || producer || consumer || drain)) {
+		fprintf(stderr, "%s: info not compatible with operating on a pcq\n\n",
+			argv[0]);
+		pcq_usage(argc, argv);
+		return -1;
+	}
+	if (create && (bucket_size == 0 || nbuckets == 0)) {
+		fprintf(stderr, "%s: create requires a bucketsize an nbuckets\n\n", argv[0]);
+		pcq_usage(argc, argv);
+		return -1;
+	}
+	if (drain && (wait || producer || nmessages || runtime)) {
+		fprintf(stderr,
+			"%s: drain can't be used with producer, time or nmessages options\n\n",
+			argv[0]);
+		pcq_usage(argc, argv);
+		return -1;
+	}
+	if (runtime && nmessages) {
+		fprintf(stderr,
+			"%s: the --nmessages and --time args cannot be used together\n\n",
+			__func__);
+		pcq_usage(argc, argv);
+		return -1;
+	}
+	if (optind > (argc - 1)) {
+		fprintf(stderr, "Must specify base filename\n\n");
+		return -1;
+	}
+	filename = argv[optind++];
+
+	if (statusfname) {
+		truncate(statusfname, 0);
+		statusfile = fopen(statusfname, "w+");
+		assert(statusfile);
+	}
+
+	if (create)
+		return pcq_create(filename, nbuckets, bucket_size, verbose);
+
+	if (info)
+		return get_queue_info(filename, statusfile, verbose);
+	if (drain) {
+		struct pcq_thread_arg ta = { 0 };
+
+		ta.role = CONSUMER;
+		ta.stop_mode = EMPTY;
+		ta.basename = filename;
+		ta.verbose = verbose;
+
+		rc = run_consumer(&ta);
+		if (ta.nerrors) {
+			if (statusfile) {
+				fprintf(statusfile, "%lld", -ta.nerrors);
+				fclose(statusfile);
+			}
+			printf("pcq: drained %lld messages from queue %s, with %lld errors\n",
+			       ta.nreceived, filename, ta.nerrors);
+			return -(int)ta.nerrors;
+		}
+
+		if (statusfile) {
+			printf("pcq: drained %lld messages from queue %s, with no errors\n",
+			       ta.nreceived, filename);
+			fprintf(statusfile, "%lld", ta.nreceived);
+			fclose(statusfile);
+		}
+		return 0;
+	}
+
+	/*
+	 * Start the producer thread if needed
+	 */
+	assert(wait);
+	if (producer) {
+		prod.role = PRODUCER;
+		prod.stop_mode = (runtime) ? STOP_FLAG : NMESSAGES;
+		prod.nmessages = nmessages;
+		prod.basename = filename;
+		prod.seed = seed;
+		prod.wait = wait;
+		prod.verbose = verbose;
+		rc = pthread_create(&producer_thread, NULL, pcq_worker, (void *)&prod);
+		if (rc) {
+			fprintf(stderr, "%s: failed to start producer thread\n", __func__);
+		}
+		sleep(0);
```

Comment:
this sleep(0) is not needed ?

Reply:
Good catch. 

That was leftover from an earlier state (when it was also non-zero), when I was creating the queue in the producer thread, and it was possible for the consumer thread to try to open the queue before it was valid. Now the protocol is to create the queue in a "pcq --create" run prior to running a producer and/or consumer.